bench: Add `--sanity-check` flag, use it in `make check`

laanwj opened the pull request:
The benchmarks are run as part of `make check` for a crash-sanity check. The actual results are being ignored. So only run them for one iteration.

This makes the `bench_bitcoin` part take 2m00 instead of 5m20 here. Which is still too long (imo), but this needs to be solved in the `WalletLoading*` benchmarks which take that long per iteration.

Also change all `bench_bitcoin` arguments to kebab-case to be consistent with the other tools (in a separate commit).

laanwj committed:
bench: Add `--sanity-check` flag, use it in `make check`

The benchmarks are run as part of `make check` for a minimum sanity
check. The actual results are being ignored. So only run them for one
iteration.

This makes the `bench_bitcoin` part take 2m00 instead of 5m20 here.
Which is still too long (imo), but this needs to be solved in the
`WalletLoading*` benchmarks which take that long per iteration.

theStack commented:
Concept ACK

laanwj commented on src/bench/bench.cpp:
```diff
@@ -69,6 +69,13 @@ void benchmark::BenchRunner::RunAll(const Args& args)
         Bench bench;
+        if (args.one_iteration) {
+            ankerl::nanobench::Config config;
+            config.mNumEpochs = 1;
```
Could also add separate flags for `--num-epochs` and `--epoch-iterations` (which default to `11` and `0` respectively).

martinus replied:
I'd use `bench.epochs(1).epochIterations(1)` instead of modifying the `Config` members directly

laanwj replied:
That's a great suggestion, will do.

laanwj commented on src/Makefile.test.include:
```diff
@@ -364,8 +364,8 @@ endif
 if ENABLE_BENCH
-	@echo "Running bench/bench_bitcoin ..."
-	$(BENCH_BINARY) > /dev/null
+	@echo "Running bench/bench_bitcoin (one iteration sanity check)..."
```
Also added a comment to preempt "why are we running the benchmarks in make check at all" questions.

jonatack commented on src/bench/bench_bitcoin.cpp:
```diff
@@ -29,6 +29,7 @@ static void SetupBenchArgs(ArgsManager& argsman)
     argsman.AddArg("-min_time=<milliseconds>", strprintf("Minimum runtime per benchmark, in milliseconds (default: %d)", DEFAULT_MIN_TIME_MS), ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION, OptionsCategory::OPTIONS);
     argsman.AddArg("-output_csv=<output.csv>", "Generate CSV file with the most important benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-output_json=<output.json>", "Generate JSON file with all benchmark results", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-one-iteration", "Run benchmarks for only one iteration (for testing purposes)", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
nit, sort here and below

jonatack replied:
The other args are snakecased and this is kebabcase; maybe good to use one style or the other.

laanwj replied:
Yes, I agree. I'll convert all the arguments to kebab-case (which is customary for UNIX-style arguments) in another commit.

jonatack commented on src/bench/bench.h:
```diff
@@ -48,6 +48,7 @@ struct Args {
     fs::path output_csv;
     fs::path output_json;
     std::string regex_filter;
+    bool one_iteration;
```
nit, place/pack this just below the first bool member

laanwj replied:
ok, sure, why not, sorting the fields by type makes as much sense as anything

jonatack replied:
yes, they've been sorted by increasing size

jonatack reviewed (commented):
Concept and approach ACK, now building and testing.

laanwj committed:
bench: Make all arguments -kebab-case

This is customary for UNIX-style arguments, and more consistent with our
other tools

laanwj commented:
Re-pushed to address @jonatack's comments.

jonatack commented on src/bench/bench.cpp:
```diff
@@ -69,6 +69,9 @@ void benchmark::BenchRunner::RunAll(const Args& args)
         Bench bench;
+        if (args.one_iteration) {
+            bench.epochs(1).epochIterations(1);
+        }
```
maybe provide user feedback that the option is running
```diff
--- a/src/bench/bench.cpp
+++ b/src/bench/bench.cpp
@@ -58,6 +58,9 @@ void benchmark::BenchRunner::RunAll(const Args& args)
     std::smatch baseMatch;
 
     std::vector<ankerl::nanobench::Result> benchmarkResults;
+    if (args.one_iteration) {
+        std::cout << "Running with --one-iteration option, i.e. epochs(1).epochIterations(1)" << std::endl;
+    }
     for (const auto& p : benchmarks()) {
         if (!std::regex_match(p.first, baseMatch, reFilter)) {
```
output
```
$ time ./src/bench/bench_bitcoin -filter=WalletBalance*.* --one-iteration
Running with --one-iteration option, i.e. epochs(1).epochIterations(1)
Warning, results might be unstable:
* DEBUG defined
* CPU frequency scaling enabled: CPU 0 between 2,485.0 and 3,100.0 MHz
* Turbo is enabled, CPU frequency will fluctuate

Recommendations
* Make sure you compile for Release
* Use 'pyperf system tune' before benchmarking. See https://github.com/psf/pyperf

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
```

sipa replied:
Revealing the fact that it's using `epochs(1).epochIterations(1)` seems like an unnecessary implementation detail for the `bench_bitcoin` user.

martinus replied:
I'd also prefer a config name like `--santiy-check`. Maybe in future this could do more to be quicker, like run all the benchmarks in parallel (I had a quick look if that works but got a segfault, probably due to some global states)

laanwj replied:
> Revealing the fact that it's using epochs(1).epochIterations(1) seems like an unnecessary implementation detail for the bench_bitcoin user.

Yeah. That's too much detail. But adding a warning message that it's active to the output makes sense, I guess.

> I'd also prefer a config name like --sanity-check. Maybe in future this could do more to be quicker, like run all the benchmarks in parallel (I had a quick look if that works but got a segfault, probably due to some global states)

Agree on the rename. Running the benchmarks in parallel would be another stress test but it seems we're not ready for that :smile: 

Though ideally I would like to aim for <10ms per iteration for all benchmarks, so running 1 iteration in parallel shouldn't have that much effect. Running 1 of everything in serial is still supposed to be fast!

fanquake replied:
> Agree on the rename.

@laanwj did you want to do that rename here? If not, I'd like to merge this soon.

laanwj replied:
Yes, I'm going to make the requested changes here and re-push in a bit.

jonatack reviewed (commented):
Tested ACK c52a71ea05c787143124b1089ef4c895f8fb01bf though the reduction in run time isn't large in my testing

```
OK
Running bench/bench_bitcoin (one iteration sanity check)...
bench/bench_bitcoin --one-iteration > /dev/null
```

Modulo typo in first commit message s/kebab-hase/kebab-case/

MarcoFalke commented:
Looks like #24924 might be related

laanwj commented:
Re-pushed:
- Argument is now `-sanity-check`.
- Add warning that results are useless.
- Typo in commit message.

hebasto commented:
Concept ACK.

hebasto commented on src/bench/bench.cpp:
```diff
@@ -57,6 +57,10 @@ void benchmark::BenchRunner::RunAll(const Args& args)
     std::smatch baseMatch;
 
+    if (args.sanity_check) {
+        std::cout << "Running with --sanity check option, benchmark results will be useless." << std::endl;
```
```suggestion
        std::cout << "Running with -sanity-check option, benchmark results will be useless." << std::endl;
```

Also s/--sanity-check/-sanity-check/ in the commit message?

jonatack replied:
I'm not sure as there isn't a single-letter option like -s here,  but per https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html: "Long options consist of -- followed by a name made of alphanumeric characters and dashes. Option names are typically one to three words long, with hyphens to separate words. Users can abbreviate the option names as long as the abbreviations are unique."

jonatack replied:
Looks like "--sanity check" in this line should be "sanity check" or "--sanity-check".

jonatack replied:
Just a missing hyphen s/--sanity check/--sanity-check/

hebasto commented on src/Makefile.test.include:
```diff
@@ -364,8 +364,8 @@ endif
-	@echo "Running bench/bench_bitcoin ..."
-	$(BENCH_BINARY) > /dev/null
+	@echo "Running bench/bench_bitcoin (one iteration sanity check)..."
+	$(BENCH_BINARY) --sanity-check > /dev/null
```
```suggestion
	$(BENCH_BINARY) -sanity-check > /dev/null
```

laanwj commented:
Maybe I'm an old UNIX fart in some things but I prefer long options prefixed with two `-`, I prefer not to change this.

hebasto reviewed (approved):
ACK 4f31c21b7f384e50e0af8275e831085d1d69b386, tested on Ubuntu 22.04.
